catchup/service_test.go

```diff
@@ -317,6 +317,13 @@ func TestAbruptWrites(t *testing.T) {
 	s.testStart()
 
 	s.sync()
+
+	for try:=0; try < 100; try++ {
+		if remote.LastRound() == local.LastRound() {
+			break
+		}
+		time.Sleep(10*time.Millisecond)
+	}
```

Comment:
Under what condition does `s.sync()` fail to complete the synchronization before returning ?

Reply:
Ledger `local` is getting blocks added to it through `sync()` and through the for loop in `go func()` above in the test.

When a block is written to the ledger via `mockedLedger AddBlock` above (service_test.go line 311), it closes the wait chan for that round, which triggers the context cancel function in service and `fetchAndWrite` gets "context canceled" err in `innerFetch`. 

Eventually, `fetchAndWrite` reaches the `catchupRetryLimit` for the round and returns. When `fetchAndWrite` returns `false`, `pipelineCallback` returns 0, which eventually leads to round 0 returned from `completed` chan in `pipelinedFetch` loop, and returns. This ends the `sync()`. 

When `sync()` returns, it happens that the for loop above is still adding blocks to the local ledger. This wait is for that to finish. 

To properly fix this, I need to understand the expected behavior of the test. Why two different routines are writing to the ledger simultaneously, and which eventually ends up the go func in the test writing all the blocks.

Reply:
It emulates that fact that the agreement can also generate new rounds.

Reply:
I think that in this case, you'd want to have the `wg.Wait()` right after the `s.sync`